bot/cogs/moderation/incidents.py

```diff
@@ -0,0 +1,331 @@
+                username=incident.author.name,
+                avatar_url=incident.author.avatar_url,
+                wait=True,  # This makes the method return the sent Message object
+            )
+
+            # Finally add the `outcome` emoji
+            await message.add_reaction(outcome.value)
+
+        except Exception as exc:
+            log.exception("Failed to archive incident to #incidents-archive", exc_info=exc)
+            return False
+
+        else:
+            log.trace("Message archived successfully!")
+            return True
+
+    def make_confirmation_task(self, incident: discord.Message, timeout: int = 5) -> asyncio.Task:
+        """
+        Create a task to wait `timeout` seconds for `incident` to be deleted.
+
+        If `timeout` passes, this will raise `asyncio.TimeoutError`, signaling that we haven't
+        been able to confirm that the message was deleted.
+        """
+        log.trace(f"Confirmation task will wait {timeout=} seconds for {incident.id=} to be deleted")
+
+        def check(payload: discord.RawReactionActionEvent) -> bool:
+            return payload.message_id == incident.id
+
+        coroutine = self.bot.wait_for(event="raw_message_delete", check=check, timeout=timeout)
+        return self.bot.loop.create_task(coroutine)
+
+    async def process_event(self, reaction: str, incident: discord.Message, member: discord.Member) -> None:
+        """
+        Process a `reaction_add` event in #incidents.
+
+        First, we check that the reaction is a recognized `Signal` member, and that it was sent by
+        a permitted user (at least one role in `ALLOWED_ROLES`). If not, the reaction is removed.
+
+        If the reaction was either `Signal.ACTIONED` or `Signal.NOT_ACTIONED`, we attempt to relay
+        the report to #incidents-archive. If successful, the original message is deleted.
+
+        We do not release `event_lock` until we receive the corresponding `message_delete` event.
+        This ensures that if there is a racing event awaiting the lock, it will fail to find the
+        message, and will abort. There is a `timeout` to ensure that this doesn't hold the lock
+        forever should something go wrong.
+        """
+        members_roles: t.Set[int] = {role.id for role in member.roles}
+        if not members_roles & ALLOWED_ROLES:  # Intersection is truthy on at least 1 common element
+            log.debug(f"Removing invalid reaction: user {member} is not permitted to send signals")
+            await incident.remove_reaction(reaction, member)
+            return
+
+        if reaction not in ALLOWED_EMOJI:
+            log.debug(f"Removing invalid reaction: emoji {reaction} is not a valid signal")
+            await incident.remove_reaction(reaction, member)
+            return
+
+        # If we reach this point, we know that `emoji` is a `Signal` member
+        signal = Signal(reaction)
+        log.trace(f"Received signal: {signal}")
```

Comment:
What do you think about using a try-except instead? "Easier to ask for forgiveness" and whatnot.

Reply:
I tried it and ended up being very happy with it. The idea crossed my mind at one point but then I ended up not doing it, because I originally had both the role and emoji check as a single conditional - they both remove the react and abort. Eventually though I felt that it'd be nice to actually log *why* it was removed, and I ended up making them disjoint.

With this change, the `ALLOWED_EMOJI` variable is only used by `has_signals` so the name no longer makes much sense, and so I ended up adjusting that as well. See: 98b8947.